src/utils/validations/component/validations/with-validation.hoc.js

```diff
@@ -0,0 +1,159 @@
+    }
+
+    static contextTypes = {
+      addInput: PropTypes.func,
+      removeInput: PropTypes.func,
+      adjustCount: PropTypes.func
+    }
+
+    state = {
+      hasError: false,
+      hasWarning: false,
+      hasInfo: false
+      // value: ''
+    };
+
+    componentDidMount = () => {
+      if (this.context && this.context.addInput && this.props.validations.length) {
+        this.context.addInput(this.props.name, this.validate);
+      }
+    }
+
+    componentWillUnmount = () => {
+      if (this.context && this.context.removeInput && this.props.validations.length) {
+        this.context.removeInput(this.props.name);
+      }
+    }
+
+    validate = async (types = ['validations', 'warnings', 'info']) => {
+      let result = true;
+      if (types.includes('validations')) result = await this.runValidation('validations');
+      if (result && types.includes('warnings')) result = await this.runValidation('warnings');
+      if (result && types.includes('info')) result = await this.runValidation('info');
+      return result;
+    }
+
+    runValidation = (type) => {
+      return new Promise((resolve) => {
+        if (this.props[type].length === 0) {
+          resolve(true);
+        }
+        const validate = validator(this.props[type]);
+        validate(this.props.value)
+          .then(() => {
+            switch (type) {
+              case 'validations':
+                if (this.state.hasError !== false && this.context.adjustCount) {
+                  this.setState({ hasError: false });
+                  this.context.adjustCount('error', -1);
+                }
+                break;
+              case 'warnings':
+                if (this.state.hasWarning !== false && this.context.adjustCount) {
+                  this.setState({ hasWarning: false });
+                  this.context.adjustCount('warning', -1);
+                }
+                break;
+              case 'info':
+                if (this.state.hasInfo !== false && this.context.adjustCount) {
+                  this.setState({ hasInfo: false });
+                  this.context.adjustCount('info', -1);
+                }
+                break;
+              default:
+            }
+            resolve(true);
+          })
+          .catch((e) => {
+            switch (type) {
+              case 'validations':
+                this.setState({ hasError: e });
+                if (this.context.adjustCount) this.context.adjustCount('error', 1);
+                break;
+              case 'warnings':
+                this.setState({ hasWarning: e });
+                if (this.context.adjustCount) this.context.adjustCount('warning', 1);
+                break;
+              case 'info':
+                this.setState({ hasInfo: e });
+                if (this.context.adjustCount) this.context.adjustCount('info', 1);
+                break;
+              default:
+            }
+            resolve(false);
+          });
+      });
+    }
```

Comment:
That method is huge and contains multiple nested blocks. Could you flatten it by splitting it into separate methods?

Reply:
I've split it up into some smaller functions, I noticed an edge case whilst doing it so had to add some further implementation to prevent the same error being counted multiple times etc